modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductVirtualSettingsFileEntryResourceImpl.java

```diff
@@ -0,0 +1,316 @@
+
+		CPDefinition cpDefinition = _cpDefinitionLocalService.getCPDefinition(
+			cpDefinitionVirtualSetting.getClassPK());
+
+		long fileEntryId = cpdVirtualSettingFileEntry.getFileEntryId();
+
+		BinaryFile binaryFile = multipartBody.getBinaryFile("file");
+
+		if (binaryFile != null) {
+			FileEntry fileEntry = FileEntryUtil.addFileEntry(
+				binaryFile, cpDefinition.getGroupId(),
+				_cpdVirtualSettingFileEntryService, _dlAppService,
+				_repositoryLocalService, _uniqueFileNameProvider,
+				_serviceContextHelper.getServiceContext(
+					cpDefinition.getGroupId()));
+
+			fileEntryId = fileEntry.getFileEntryId();
+		}
+		else if (productVirtualSettingsFileEntry.getAttachment() != null) {
+			fileEntryId = FileEntryUtil.getFileEntryId(
+				productVirtualSettingsFileEntry.getAttachment(),
+				productVirtualSettingsFileEntry.getUrl(),
+				_uniqueFileNameProvider,
+				_serviceContextHelper.getServiceContext(
+					cpDefinition.getGroupId()));
+		}
+
+		return _toProductVirtualSettingsFileEntry(
+			_cpdVirtualSettingFileEntryService.updateCPDefinitionVirtualSetting(
+				id, fileEntryId,
+				GetterUtil.get(
+					productVirtualSettingsFileEntry.getUrl(),
+					cpdVirtualSettingFileEntry.getUrl()),
+				GetterUtil.get(
+					productVirtualSettingsFileEntry.getVersion(),
+					cpdVirtualSettingFileEntry.getVersion())));
+	}
+
+	@Override
+	public ProductVirtualSettingsFileEntry
+			postProductVirtualSettingIdProductVirtualSettingsFileEntry(
+				Long id, MultipartBody multipartBody)
+		throws Exception {
+
+		ProductVirtualSettingsFileEntry productVirtualSettingsFileEntry =
+			multipartBody.getValueAsNullableInstance(
+				"productVirtualSettingsFileEntry",
+				ProductVirtualSettingsFileEntry.class);
+
+		BinaryFile binaryFile = multipartBody.getBinaryFile("file");
+
+		if ((binaryFile == null) &&
+			(productVirtualSettingsFileEntry.getAttachment() == null)) {
+
+			throw new BadRequestException("No file found in body");
+		}
+
+		CPDefinitionVirtualSetting cpDefinitionVirtualSetting =
+			_cpDefinitionVirtualSettingService.getCPDefinitionVirtualSetting(
+				id);
+
+		CPDefinition cpDefinition = _cpDefinitionLocalService.getCPDefinition(
+			cpDefinitionVirtualSetting.getClassPK());
+
+		long fileEntryId = 0;
+
+		if (binaryFile != null) {
+			FileEntry fileEntry = FileEntryUtil.addFileEntry(
+				binaryFile, cpDefinition.getGroupId(),
+				_cpdVirtualSettingFileEntryService, _dlAppService,
+				_repositoryLocalService, _uniqueFileNameProvider,
+				_serviceContextHelper.getServiceContext(
+					cpDefinition.getGroupId()));
+
+			fileEntryId = fileEntry.getFileEntryId();
+		}
+		else if (productVirtualSettingsFileEntry.getAttachment() != null) {
+			fileEntryId = FileEntryUtil.getFileEntryId(
+				productVirtualSettingsFileEntry.getAttachment(),
+				productVirtualSettingsFileEntry.getUrl(),
+				_uniqueFileNameProvider,
+				_serviceContextHelper.getServiceContext(
+					cpDefinition.getGroupId()));
+		}
```

Comment:
this code is very similar to the one in the patch... is it possible to extract it to a private method?